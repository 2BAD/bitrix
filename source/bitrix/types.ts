import { Diff } from './utils/Diff'

// @todo We'd want to have gettable and listable as separate enums,
//       but we can't due too types compatibility issues when `list` uses `get` internally
export enum APIMethod {
  // Gettable
  BATCH = 'batch',
  GET_DEAL = 'crm.deal.get',
  GET_LEAD = 'crm.lead.get',

  // Listable
  LIST_DEALS = 'crm.deal.list',
  LIST_LEADS = 'crm.lead.list'
}

const BITRIX_LISTABLE_METHODS = [APIMethod.LIST_DEALS, APIMethod.LIST_LEADS] as const

export type BitrixListableMethod = typeof BITRIX_LISTABLE_METHODS[number]
export type BitrixGettableMethod = Diff<APIMethod, BitrixListableMethod>

export interface BitrixPayloadTime {
  readonly start: number
  readonly finish: number
  readonly duration: number
  readonly processing: number
  readonly date_start: string
  readonly date_finish: string
}

export interface BitrixGetPayload<P> {
  readonly result: P,
  readonly time: BitrixPayloadTime
}

export interface BitrixListPayload<P> {
  readonly result: readonly P[],
  readonly error?: string,
  readonly total: number,
  readonly next?: number
  readonly time: BitrixPayloadTime
}

// `C` stands for a map of names to structural types they will uphold in result
// `[]` in language of ill Bitrix means `undefined`. Just deal with it.
// Also, it will return object if command names specified and array if names are numbers. Deal with it.
export interface BitrixBatchPayload<C> {
  readonly result: {
    readonly result: { readonly [P in keyof C]?: C[P] } | ReadonlyArray<C[keyof C]>
    readonly result_error: { readonly [P in keyof C]?: string } | readonly string[]
    readonly result_total: { readonly [P in keyof C]?: number } | readonly number[]
    readonly result_next: { readonly [P in keyof C]?: number } | readonly number[]
    readonly result_time: { readonly [P in keyof C]?: BitrixPayloadTime } | readonly BitrixPayloadTime[]
  }
  readonly time: BitrixPayloadTime
}

export interface BitrixCommandParams {
  readonly [key: string]: string | number
}

export interface BitrixCommand {
  readonly method: APIMethod
  readonly params?: BitrixCommandParams
}

export interface BitrixCommands {
  readonly [key: string]: BitrixCommand
}

export interface BitrixListOptions {
  readonly start?: number
}

// =====================================================
// Bitrix Payloads
// =====================================================

// String which is actually a number, like `'20.23'`
export type NumberString = string
// Like `'2018-06-07T03:00:00+03:00'`
export type ISODate = string
export type BitrixBoolString = 'Y' | 'N'

// @todo This is approximate structure. Might not be accurate
export interface BitrixDeal {
  readonly ID: NumberString
  readonly TITLE: string
  readonly TYPE_ID: string
  readonly STAGE_ID: string
  readonly PROBABILITY: unknown // @todo Check is it right
  readonly CURRENCY_ID: string
  readonly OPPORTUNITY: NumberString
  readonly TAX_VALUE: NumberString
  readonly LEAD_ID: NumberString
  readonly COMPANY_ID: NumberString
  readonly CONTACT_ID: NumberString | null
  readonly QUOTE_ID: NumberString | null
  readonly BEGINDATE: ISODate
  readonly CLOSEDATE: ISODate
  readonly ASSIGNED_BY_ID: NumberString
  readonly CREATED_BY_ID: NumberString
  readonly MODIFY_BY_ID: NumberString
  readonly DATE_CREATE: ISODate
  readonly DATE_MODIFY: ISODate
  readonly OPENED: BitrixBoolString
  readonly CLOSED: BitrixBoolString
  readonly COMMENTS: string
  readonly ADDITIONAL_INFO: unknown // @todo Check is it right
  readonly LOCATION_ID: NumberString | null
  readonly CATEGORY_ID: NumberString
  readonly STAGE_SEMANTIC_ID: NumberString
  readonly IS_NEW: BitrixBoolString
  readonly IS_RECURRING: BitrixBoolString
  readonly IS_RETURN_CUSTOMER: BitrixBoolString
  readonly IS_REPEATED_APPROACH: BitrixBoolString
  readonly SOURCE_ID: NumberString | null
  readonly SOURCE_DESCRIPTION: string | null
  readonly ORIGINATOR_ID: string
  readonly ORIGIN_ID: NumberString
  readonly UTM_SOURCE: string | null
  readonly UTM_MEDIUM: string | null
  readonly UTM_CAMPAIGN: string | null
  readonly UTM_CONTENT: string | null
  readonly UTM_TERM: string | null
}
// @todo This is approximate structure. Might not be accurate
export interface BitrixLead {
  readonly ID: NumberString
  readonly TITLE: string
  readonly HONORIFIC: unknown | null // @todo
  readonly NAME: string | null
  readonly SECOND_NAME: string | null
  readonly LAST_NAME: string | null
  readonly COMPANY_TITLE: string | null
  readonly COMPANY_ID: unknown | null // @todo
  readonly CONTACT_ID: unknown | null // @todo
  readonly IS_RETURN_CUSTOMER: BitrixBoolString
  readonly BIRTHDATE: unknown | null // @todo
  readonly SOURCE_ID: NumberString
  readonly SOURCE_DESCRIPTION: null
  readonly STATUS_ID: unknown | null // @todo
  readonly STATUS_DESCRIPTION: string | null
  readonly POST: unknown | null // @todo
  readonly COMMENTS: string | null
  readonly CURRENCY_ID: string
  readonly OPPORTUNITY: NumberString
  readonly HAS_PHONE: BitrixBoolString
  readonly HAS_EMAIL: BitrixBoolString
  readonly HAS_IMOL: BitrixBoolString
  readonly ASSIGNED_BY_ID: NumberString
  readonly CREATED_BY_ID: NumberString
  readonly MODIFY_BY_ID: NumberString
  readonly DATE_CREATE: ISODate
  readonly DATE_MODIFY: ISODate
  readonly DATE_CLOSED: ISODate
  readonly STATUS_SEMANTIC_ID: string
  readonly OPENED: BitrixBoolString
  readonly ORIGINATOR_ID: unknown | null // @todo
  readonly ORIGIN_ID: unknown | null // @todo
  readonly ADDRESS: string | null
  readonly ADDRESS_2: string | null
  readonly ADDRESS_CITY: string | null
  readonly ADDRESS_POSTAL_CODE: string | null
  readonly ADDRESS_REGION: string | null
  readonly ADDRESS_PROVINCE: string | null
  readonly ADDRESS_COUNTRY: string | null
  readonly ADDRESS_COUNTRY_CODE: string | null
  readonly UTM_SOURCE: string | null
  readonly UTM_MEDIUM: string | null
  readonly UTM_CAMPAIGN: string | null
  readonly UTM_CONTENT: string | null
  readonly UTM_TERM: string | null
}
