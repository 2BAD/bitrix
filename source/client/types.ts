import { Diff } from '../utils/Diff'

export enum Method {
  // Gettable
  BATCH = 'batch',

  GET_CONTACT = 'crm.contact.get',
  GET_DEAL = 'crm.deal.get',
  GET_LEAD = 'crm.lead.get',
  GET_STATUS = 'crm.status.get',
  GET_USER = 'user.get',

  CREATE_CONTACT = 'crm.contact.add',
  CREATE_DEAL = 'crm.deal.add',
  CREATE_LEAD = 'crm.lead.add',
  CREATE_STATUS = 'crm.status.add',

  UPDATE_CONTACT = 'crm.contact.update',
  UPDATE_DEAL = 'crm.deal.update',
  UPDATE_LEAD = 'crm.lead.update',
  UPDATE_STATUS = 'crm.status.update',

  // Listable
  LIST_CONTACTS = 'crm.contact.list',
  LIST_DEALS = 'crm.deal.list',
  LIST_LEADS = 'crm.lead.list',
  LIST_STATUSES = 'crm.status.list',
  // yes, this one is correct, they don't have separate `list` method and this one returns all users
  LIST_USERS = 'user.search'
}

const LISTABLE_METHODS = [
  Method.LIST_CONTACTS,
  Method.LIST_DEALS,
  Method.LIST_LEADS,
  Method.LIST_STATUSES,
  Method.LIST_USERS
] as const

export type ListableMethod = typeof LISTABLE_METHODS[number]
export type GettableMethod = Diff<Method, ListableMethod>

export interface PayloadTime {
  readonly start: number
  readonly finish: number
  readonly duration: number
  readonly processing: number
  readonly date_start: string
  readonly date_finish: string
}

export interface GetPayload<P> {
  readonly result: P,
  readonly time: PayloadTime
}

export interface ListPayload<P> {
  readonly result: readonly P[],
  readonly error?: string,
  readonly total: number,
  readonly next?: number
  readonly time: PayloadTime
}

// `C` stands for a map of names to structural types they will uphold in result
// `[]` in language of ill Bitrix means `undefined`. Just deal with it.
// Also, it will return object if command names specified and array if names are numbers. Deal with it.
export interface BatchPayload<C> {
  readonly result: {
    readonly result: { readonly [P in keyof C]?: C[P] } | ReadonlyArray<C[keyof C]>
    readonly result_error: { readonly [P in keyof C]?: string } | readonly string[]
    readonly result_total: { readonly [P in keyof C]?: number } | readonly number[]
    readonly result_next: { readonly [P in keyof C]?: number } | readonly number[]
    readonly result_time: { readonly [P in keyof C]?: PayloadTime } | readonly PayloadTime[]
  }
  readonly time: PayloadTime
}

export type Payload<P> = GetPayload<P> | ListPayload<P> | BatchPayload<P>

// @todo Figure out full list of possible values
export interface ListParams {
  readonly start?: number
  readonly order?: { readonly [key: string]: 'ASC' }
  readonly filter?: { readonly '>PROBABILITY': number }
  readonly select?: ReadonlyArray<'*' | 'UF_*' | string>
}

export interface Command {
  readonly method: Method
  readonly params?: Record<string, any>
}

export type Commands =
  { readonly [key: string]: Command } |
  // For arrays. It's signature, since `Command[]` won't be
  // accepted by types like `Record`
  { readonly [index: number]: Command }
