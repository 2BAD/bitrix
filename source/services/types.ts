// =====================================================
// Bitrix Payloads
// =====================================================

// String which is actually a number, like `'20.23'`
export type NumberString = string
// Like `'2018-06-07T03:00:00+03:00'`
export type ISODate = string
export type BoolString = 'Y' | 'N'

// @todo This is approximate structure. Might not be accurate
export interface Deal {
  // Deal can have user fields
  // @todo Remove `unknown` when all fields will be known
  readonly [key: string]: string | unknown

  // tslint:disable-next-line: no-mixed-interface
  readonly ID: NumberString
  readonly TITLE: string
  readonly TYPE_ID: string
  readonly STAGE_ID: string
  readonly PROBABILITY: unknown // @todo Check is it right
  readonly CURRENCY_ID: string
  readonly OPPORTUNITY: NumberString
  readonly TAX_VALUE: NumberString
  readonly LEAD_ID: NumberString
  readonly COMPANY_ID: NumberString
  readonly CONTACT_ID: NumberString | null
  readonly QUOTE_ID: NumberString | null
  readonly BEGINDATE: ISODate
  readonly CLOSEDATE: ISODate
  readonly ASSIGNED_BY_ID: NumberString
  readonly CREATED_BY_ID: NumberString
  readonly MODIFY_BY_ID: NumberString
  readonly DATE_CREATE: ISODate
  readonly DATE_MODIFY: ISODate
  readonly OPENED: BoolString
  readonly CLOSED: BoolString
  readonly COMMENTS: string
  readonly ADDITIONAL_INFO: unknown // @todo Check is it right
  readonly LOCATION_ID: NumberString | null
  readonly CATEGORY_ID: NumberString
  readonly STAGE_SEMANTIC_ID: NumberString
  readonly IS_NEW: BoolString
  readonly IS_RECURRING: BoolString
  readonly IS_RETURN_CUSTOMER: BoolString
  readonly IS_REPEATED_APPROACH: BoolString
  readonly SOURCE_ID: NumberString | null
  readonly SOURCE_DESCRIPTION: string | null
  readonly ORIGINATOR_ID: string
  readonly ORIGIN_ID: NumberString
  readonly UTM_SOURCE: string | null
  readonly UTM_MEDIUM: string | null
  readonly UTM_CAMPAIGN: string | null
  readonly UTM_CONTENT: string | null
  readonly UTM_TERM: string | null
}

// @todo This is approximate structure. Might not be accurate
export interface Lead {
  // Deal can have user fields
  // @todo Remove `unknown` when all fields will be known
  readonly [key: string]: string | unknown

  // tslint:disable-next-line: no-mixed-interface
  readonly ID: NumberString
  readonly TITLE: string
  readonly HONORIFIC: unknown | null // @todo
  readonly NAME: string | null
  readonly SECOND_NAME: string | null
  readonly LAST_NAME: string | null
  readonly COMPANY_TITLE: string | null
  readonly COMPANY_ID: unknown | null // @todo
  readonly CONTACT_ID: unknown | null // @todo
  readonly IS_RETURN_CUSTOMER: BoolString
  readonly BIRTHDATE: unknown | null // @todo
  readonly SOURCE_ID: NumberString
  readonly SOURCE_DESCRIPTION: null
  readonly STATUS_ID: unknown | null // @todo
  readonly STATUS_DESCRIPTION: string | null
  readonly POST: unknown | null // @todo
  readonly COMMENTS: string | null
  readonly CURRENCY_ID: string
  readonly OPPORTUNITY: NumberString
  readonly HAS_PHONE: BoolString
  readonly HAS_EMAIL: BoolString
  readonly HAS_IMOL: BoolString
  readonly ASSIGNED_BY_ID: NumberString
  readonly CREATED_BY_ID: NumberString
  readonly MODIFY_BY_ID: NumberString
  readonly DATE_CREATE: ISODate
  readonly DATE_MODIFY: ISODate
  readonly DATE_CLOSED: ISODate
  readonly STATUS_SEMANTIC_ID: string
  readonly OPENED: BoolString
  readonly ORIGINATOR_ID: unknown | null // @todo
  readonly ORIGIN_ID: unknown | null // @todo
  readonly ADDRESS: string | null
  readonly ADDRESS_2: string | null
  readonly ADDRESS_CITY: string | null
  readonly ADDRESS_POSTAL_CODE: string | null
  readonly ADDRESS_REGION: string | null
  readonly ADDRESS_PROVINCE: string | null
  readonly ADDRESS_COUNTRY: string | null
  readonly ADDRESS_COUNTRY_CODE: string | null
  readonly UTM_SOURCE: string | null
  readonly UTM_MEDIUM: string | null
  readonly UTM_CAMPAIGN: string | null
  readonly UTM_CONTENT: string | null
  readonly UTM_TERM: string | null
}

export interface StatusExtra {
  readonly SEMANTICS: string
  readonly COLOR: string
}

export interface Status {
  readonly ID: NumberString
  readonly ENTITY_ID: string
  readonly STATUS_ID: NumberString
  readonly NAME: string
  readonly NAME_INIT: string
  readonly SORT: NumberString
  readonly SYSTEM: BoolString
  readonly EXTRA: StatusExtra | undefined
}
