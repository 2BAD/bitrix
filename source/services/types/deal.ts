import { BoolString, ISODate, NumberString } from './common'

export interface Deal {
  // Deal can have user fields
  readonly [key: string]: string | null

  // tslint:disable-next-line: no-mixed-interface
  readonly ID: NumberString
  readonly TITLE: string
  readonly TYPE_ID: string
  readonly STAGE_ID: string
  readonly PROBABILITY: NumberString | null
  readonly CURRENCY_ID: string
  readonly OPPORTUNITY: NumberString
  readonly TAX_VALUE: NumberString
  readonly LEAD_ID: NumberString
  readonly COMPANY_ID: NumberString
  readonly CONTACT_ID: NumberString | null
  readonly QUOTE_ID: NumberString | null
  readonly BEGINDATE: ISODate
  readonly CLOSEDATE: ISODate
  readonly ASSIGNED_BY_ID: NumberString
  readonly CREATED_BY_ID: NumberString
  readonly MODIFY_BY_ID: NumberString
  readonly DATE_CREATE: ISODate
  readonly DATE_MODIFY: ISODate
  readonly OPENED: BoolString
  readonly CLOSED: BoolString
  readonly COMMENTS: string
  readonly ADDITIONAL_INFO: NumberString | null
  readonly LOCATION_ID: NumberString | null
  readonly CATEGORY_ID: NumberString
  readonly STAGE_SEMANTIC_ID: NumberString
  readonly IS_NEW: BoolString
  readonly IS_RECURRING: BoolString
  readonly IS_RETURN_CUSTOMER: BoolString
  readonly IS_REPEATED_APPROACH: BoolString
  readonly SOURCE_ID: NumberString | null
  readonly SOURCE_DESCRIPTION: string | null
  readonly ORIGINATOR_ID: string
  readonly ORIGIN_ID: NumberString
  readonly UTM_SOURCE: string | null
  readonly UTM_MEDIUM: string | null
  readonly UTM_CAMPAIGN: string | null
  readonly UTM_CONTENT: string | null
  readonly UTM_TERM: string | null
}
