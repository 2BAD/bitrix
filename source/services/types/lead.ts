import { BoolString, ISODate, NumberString } from './common'

// @todo This is approximate structure. Might not be accurate
export interface Lead {
  // Deal can have user fields
  // @todo Remove `unknown` when all fields will be known
  readonly [key: string]: string | unknown

  // tslint:disable-next-line: no-mixed-interface
  readonly ID: NumberString
  readonly TITLE: string
  readonly HONORIFIC: unknown | null // @todo
  readonly NAME: string | null
  readonly SECOND_NAME: string | null
  readonly LAST_NAME: string | null
  readonly COMPANY_TITLE: string | null
  readonly COMPANY_ID: unknown | null // @todo
  readonly CONTACT_ID: unknown | null // @todo
  readonly IS_RETURN_CUSTOMER: BoolString
  readonly BIRTHDATE: unknown | null // @todo
  readonly SOURCE_ID: NumberString
  readonly SOURCE_DESCRIPTION: null
  readonly STATUS_ID: unknown | null // @todo
  readonly STATUS_DESCRIPTION: string | null
  readonly POST: unknown | null // @todo
  readonly COMMENTS: string | null
  readonly CURRENCY_ID: string
  readonly OPPORTUNITY: NumberString
  readonly HAS_PHONE: BoolString
  readonly HAS_EMAIL: BoolString
  readonly HAS_IMOL: BoolString
  readonly ASSIGNED_BY_ID: NumberString
  readonly CREATED_BY_ID: NumberString
  readonly MODIFY_BY_ID: NumberString
  readonly DATE_CREATE: ISODate
  readonly DATE_MODIFY: ISODate
  readonly DATE_CLOSED: ISODate
  readonly STATUS_SEMANTIC_ID: string
  readonly OPENED: BoolString
  readonly ORIGINATOR_ID: unknown | null // @todo
  readonly ORIGIN_ID: unknown | null // @todo
  readonly ADDRESS: string | null
  readonly ADDRESS_2: string | null
  readonly ADDRESS_CITY: string | null
  readonly ADDRESS_POSTAL_CODE: string | null
  readonly ADDRESS_REGION: string | null
  readonly ADDRESS_PROVINCE: string | null
  readonly ADDRESS_COUNTRY: string | null
  readonly ADDRESS_COUNTRY_CODE: string | null
  readonly UTM_SOURCE: string | null
  readonly UTM_MEDIUM: string | null
  readonly UTM_CAMPAIGN: string | null
  readonly UTM_CONTENT: string | null
  readonly UTM_TERM: string | null
}
