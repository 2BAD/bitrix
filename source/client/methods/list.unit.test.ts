/* tslint-env jest */
// tslint:disable: no-expression-statement object-literal-sort-keys no-magic-numbers

import { Method } from '../../method.types'
import List, { batchToListPayload, fillWithCommands, highest } from './list'

describe('Client `fillWithCommands` method', () => {
  it('should fill array with required amount of commands to process all entries', () => {
    const command = { method: Method.CRM_DEAL_LIST, params: { select: ['*'] } }
    const start = 0
    const toProcess = 7
    const entriesPerCommand = 2

    expect(fillWithCommands(command, start, toProcess, entriesPerCommand)).toMatchSnapshot()
  })

  it('should properly take into account start', () => {
    const command = { method: Method.CRM_DEAL_LIST, params: { select: ['*'] } }
    const start = 2
    const toProcess = 7
    const entriesPerCommand = 2

    expect(fillWithCommands(command, start, toProcess, entriesPerCommand)).toMatchSnapshot()
  })

  it('should override `params.start`', () => {
    const wrongStart = 99
    const command = { method: Method.CRM_DEAL_LIST, params: { start: wrongStart } }
    const start = 0
    const toProcess = 2
    const entriesPerCommand = 1
    const commands = fillWithCommands(command, start, toProcess, entriesPerCommand)

    expect(Object.values(commands).map((c) => c.params)).toMatchSnapshot()
  })
})

describe('Client `highest` method', () => {
  it('should get highest value from object', () => {
    expect(highest({ a: undefined })).toBe(undefined)
    expect(highest({ a: undefined, b: undefined })).toBe(undefined)
    expect(highest({ a: 1, b: undefined })).toBe(1)
    expect(highest({ a: undefined, b: 1 })).toBe(1)
    expect(highest({ a: 1, b: 2 })).toBe(2)
    expect(highest({ a: 2, b: 1 })).toBe(2)
  })

  it('should get highest value from array', () => {
    expect(highest([undefined])).toBe(undefined)
    expect(highest([undefined, undefined])).toBe(undefined)
    expect(highest([1, undefined])).toBe(1)
    expect(highest([undefined, 1])).toBe(1)
    expect(highest([1, 2])).toBe(2)
    expect(highest([2, 1])).toBe(2)
  })
})

describe('Client `batchToListPayload` method', () => {
  it('should convert batch with named payloads to a list payload', () => {
    const payload = {
      result: {
        result: {
          a: [{ ID: '1' }, { ID: '2' }],
          b: [{ ID: '3' }, { ID: '4' }]
        },
        result_error: { a: 'Expected error A', b: 'Expected error B' },
        result_total: { a: 4, b: 4 },
        result_next: { a: 2, b: 4 },
        result_time: {
          a: {
            start: 1567196891.008149,
            finish: 1567196891.022234,
            duration: 0.014085054397583008,
            processing: 0.013998985290527344,
            date_start: '2019-08-30T23:28:11+03:00',
            date_finish: '2019-08-30T23:28:11+03:00'
          },
          b: {
            start: 1567196891.022316,
            finish: 1567196891.03225,
            duration: 0.009933948516845703,
            processing: 0.009846210479736328,
            date_start: '2019-08-30T23:28:11+03:00',
            date_finish: '2019-08-30T23:28:11+03:00'
          }
        }
      },
      time: {
        start: 1567196890.959017,
        finish: 1567196891.223739,
        duration: 0.2647218704223633,
        processing: 0.21567082405090332,
        date_start: '2019-08-30T23:28:10+03:00',
        date_finish: '2019-08-30T23:28:11+03:00'
      }
    }

    expect(batchToListPayload(payload)).toMatchSnapshot()
  })

  it('should convert batch with payload of arrays to a list payload', () => {
    const payload = {
      result: {
        result: [
          [{ ID: '1' }, { ID: '2' }],
          [{ ID: '3' }, { ID: '4' }]
        ],
        result_error: ['Expected error 1', 'Expected error 2'],
        result_total: [4, 4],
        result_next: [2, 4],
        result_time: [{
          start: 1567196891.008149,
          finish: 1567196891.022234,
          duration: 0.014085054397583008,
          processing: 0.013998985290527344,
          date_start: '2019-08-30T23:28:11+03:00',
          date_finish: '2019-08-30T23:28:11+03:00'
        }, {
          start: 1567196891.022316,
          finish: 1567196891.03225,
          duration: 0.009933948516845703,
          processing: 0.009846210479736328,
          date_start: '2019-08-30T23:28:11+03:00',
          date_finish: '2019-08-30T23:28:11+03:00'
        }]
      },
      time: {
        start: 1567196890.959017,
        finish: 1567196891.223739,
        duration: 0.2647218704223633,
        processing: 0.21567082405090332,
        date_start: '2019-08-30T23:28:10+03:00',
        date_finish: '2019-08-30T23:28:11+03:00'
      }
    }

    expect(batchToListPayload(payload)).toMatchSnapshot()
  })

  it('should handle undefined in `result`', () => {
    const payload = {
      result: {
        result: [undefined, [{ ID: '1' }], undefined] as any,
        result_error: [] as any,
        result_total: [] as any,
        result_next: [] as any,
        result_time: [{
          start: 1567196891.008149,
          finish: 1567196891.022234,
          duration: 0.014085054397583008,
          processing: 0.013998985290527344,
          date_start: '2019-08-30T23:28:11+03:00',
          date_finish: '2019-08-30T23:28:11+03:00'
        }, {
          start: 1567196891.022316,
          finish: 1567196891.03225,
          duration: 0.009933948516845703,
          processing: 0.009846210479736328,
          date_start: '2019-08-30T23:28:11+03:00',
          date_finish: '2019-08-30T23:28:11+03:00'
        }]
      },
      time: {
        start: 1567196890.959017,
        finish: 1567196891.223739,
        duration: 0.2647218704223633,
        processing: 0.21567082405090332,
        date_start: '2019-08-30T23:28:10+03:00',
        date_finish: '2019-08-30T23:28:11+03:00'
      }
    }

    expect(batchToListPayload(payload)).toMatchSnapshot()
  })
})

describe('Client `list` method', () => {
  it('should make one request when entries can be fetched in one go', async () => {
    const callMock = jest.fn(() => Promise.resolve({ next: false }) as any)
    const batchMock = jest.fn(() => Promise.resolve({}) as any)
    const list = List({ call: callMock, batch: batchMock })

    await list(Method.CRM_DEAL_LIST, {})

    expect(callMock.mock.calls).toMatchSnapshot()
    expect(batchMock).not.toBeCalled()
  })

  it('should make multiple requests when entries can not be fetched in one go', async () => {
    const callMock = jest.fn(() => Promise.resolve({ next: 2, total: 120 }) as any)
    const batchMock = jest.fn(() => Promise.resolve({
      result: {
        result: ['done'],
        result_error: [],
        result_next: [],
        result_time: [],
        result_total: []
      },
      time: {}
    }) as any)
    const list = List({ call: callMock, batch: batchMock })

    await list(Method.CRM_DEAL_LIST, {})

    expect(callMock.mock.calls).toMatchSnapshot()
    expect(batchMock.mock.calls).toMatchSnapshot()
  })

  it('should return a first request payload when entities can be fetched in a single request', async () => {
    const mockPayload = { next: false, result: 'test' }
    const callMock = jest.fn(() => Promise.resolve(mockPayload) as any)
    const batchMock = jest.fn(() => Promise.resolve({}) as any)
    const list = List({ call: callMock, batch: batchMock })

    const payload = await list(Method.CRM_DEAL_LIST, {})

    expect(payload).toEqual(mockPayload)
  })

  it('should return a flatten payload when entities can not be fetched in a single request', async () => {
    const mockPayload = {
      result: {
        result: {
          a: [{ ID: '1' }, { ID: '2' }],
          b: [{ ID: '3' }, { ID: '4' }]
        },
        result_error: { a: 'Expected error A', b: 'Expected error B' },
        result_total: { a: 4, b: 4 },
        result_next: { a: 2, b: 4 },
        result_time: {
          a: { start: 1567196891.008149 },
          b: { start: 1567196891.022316 }
        }
      },
      time: { start: 1567196890.959017 }
    }

    const callMock = jest.fn(() => Promise.resolve({ next: 2 }) as any)
    const batchMock = jest.fn(() => Promise.resolve(mockPayload) as any)
    const list = List({ call: callMock, batch: batchMock })

    const payload = await list(Method.CRM_DEAL_LIST, {})

    expect(payload).toMatchSnapshot()
  })

  it('should default start to 0', async () => {
    const callMock = jest.fn(() => Promise.resolve({ next: 2, total: 20 }) as any)
    const batchMock = jest.fn(() => Promise.resolve({
      result: {
        result: ['done'],
        result_error: [],
        result_next: [],
        result_time: [],
        result_total: []
      },
      time: {}
    }) as any)
    const list = List({ call: callMock, batch: batchMock })

    await list(Method.CRM_DEAL_LIST, {})

    expect(callMock.mock.calls[0]).toMatchSnapshot()
    expect(batchMock.mock.calls[0]).toMatchSnapshot()
  })

  it('should properly take into account start', async () => {
    const callMock = jest.fn(() => Promise.resolve({ next: 2, total: 120 }) as any)
    const batchMock = jest.fn(() => Promise.resolve({
      result: {
        result: ['done'],
        result_error: [],
        result_next: [],
        result_time: [],
        result_total: []
      },
      time: {}
    }) as any)
    const list = List({ call: callMock, batch: batchMock })

    await list(Method.CRM_DEAL_LIST, { start: 27 })

    expect(callMock.mock.calls).toMatchSnapshot()
    expect(batchMock.mock.calls).toMatchSnapshot()
  })

  it.todo('should cast first request payload to the <P>')
  it.todo('should cast payload results to the <P>')
})
