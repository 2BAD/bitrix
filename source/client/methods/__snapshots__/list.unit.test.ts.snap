// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bitrix \`batchToListPayload\` method should convert batch with named payloads to a list payload 1`] = `
Object {
  "error": "Expected error A
Expected error B",
  "next": 4,
  "result": Array [
    Object {
      "ID": "1",
    },
    Object {
      "ID": "2",
    },
    Object {
      "ID": "3",
    },
    Object {
      "ID": "4",
    },
  ],
  "time": Object {
    "date_finish": "2019-08-30T23:28:11+03:00",
    "date_start": "2019-08-30T23:28:10+03:00",
    "duration": 0.2647218704223633,
    "finish": 1567196891.223739,
    "processing": 0.21567082405090332,
    "start": 1567196890.959017,
  },
  "total": 4,
}
`;

exports[`Bitrix \`batchToListPayload\` method should convert batch with payload of arrays to a list payload 1`] = `
Object {
  "error": "Expected error 1
Expected error 2",
  "next": 4,
  "result": Array [
    Object {
      "ID": "1",
    },
    Object {
      "ID": "2",
    },
    Object {
      "ID": "3",
    },
    Object {
      "ID": "4",
    },
  ],
  "time": Object {
    "date_finish": "2019-08-30T23:28:11+03:00",
    "date_start": "2019-08-30T23:28:10+03:00",
    "duration": 0.2647218704223633,
    "finish": 1567196891.223739,
    "processing": 0.21567082405090332,
    "start": 1567196890.959017,
  },
  "total": 4,
}
`;

exports[`Bitrix \`batchToListPayload\` method should handle undefined in \`result\` 1`] = `
Object {
  "error": "",
  "next": undefined,
  "result": Array [
    Object {
      "ID": "1",
    },
  ],
  "time": Object {
    "date_finish": "2019-08-30T23:28:11+03:00",
    "date_start": "2019-08-30T23:28:10+03:00",
    "duration": 0.2647218704223633,
    "finish": 1567196891.223739,
    "processing": 0.21567082405090332,
    "start": 1567196890.959017,
  },
  "total": 0,
}
`;

exports[`Bitrix \`fillWithCommands\` method should fill array with required amount of commands to process all entries 1`] = `
Array [
  Object {
    "method": "crm.deal.list",
    "params": Object {
      "select": Array [
        "*",
      ],
      "start": 0,
    },
  },
  Object {
    "method": "crm.deal.list",
    "params": Object {
      "select": Array [
        "*",
      ],
      "start": 2,
    },
  },
  Object {
    "method": "crm.deal.list",
    "params": Object {
      "select": Array [
        "*",
      ],
      "start": 4,
    },
  },
  Object {
    "method": "crm.deal.list",
    "params": Object {
      "select": Array [
        "*",
      ],
      "start": 6,
    },
  },
]
`;

exports[`Bitrix \`fillWithCommands\` method should override \`params.start\` 1`] = `
Array [
  Object {
    "start": 0,
  },
  Object {
    "start": 1,
  },
]
`;

exports[`Bitrix \`fillWithCommands\` method should properly take into account start 1`] = `
Array [
  Object {
    "method": "crm.deal.list",
    "params": Object {
      "select": Array [
        "*",
      ],
      "start": 2,
    },
  },
  Object {
    "method": "crm.deal.list",
    "params": Object {
      "select": Array [
        "*",
      ],
      "start": 4,
    },
  },
  Object {
    "method": "crm.deal.list",
    "params": Object {
      "select": Array [
        "*",
      ],
      "start": 6,
    },
  },
]
`;

exports[`Bitrix \`list\` method should default start to 0 1`] = `
Array [
  "crm.deal.list",
  Object {
    "start": 0,
  },
]
`;

exports[`Bitrix \`list\` method should default start to 0 2`] = `
Array [
  Array [
    Object {
      "method": "crm.deal.list",
      "params": Object {
        "start": 0,
      },
    },
  ],
]
`;

exports[`Bitrix \`list\` method should make multiple requests when entries can not be fetched in one go 1`] = `
Array [
  Array [
    "crm.deal.list",
    Object {
      "start": 0,
    },
  ],
]
`;

exports[`Bitrix \`list\` method should make multiple requests when entries can not be fetched in one go 2`] = `
Array [
  Array [
    Array [
      Object {
        "method": "crm.deal.list",
        "params": Object {
          "start": 0,
        },
      },
      Object {
        "method": "crm.deal.list",
        "params": Object {
          "start": 50,
        },
      },
      Object {
        "method": "crm.deal.list",
        "params": Object {
          "start": 100,
        },
      },
    ],
  ],
]
`;

exports[`Bitrix \`list\` method should make one request when entries can be fetched in one go 1`] = `
Array [
  Array [
    "crm.deal.list",
    Object {
      "start": 0,
    },
  ],
]
`;

exports[`Bitrix \`list\` method should properly take into account start 1`] = `
Array [
  Array [
    "crm.deal.list",
    Object {
      "start": 27,
    },
  ],
]
`;

exports[`Bitrix \`list\` method should properly take into account start 2`] = `
Array [
  Array [
    Array [
      Object {
        "method": "crm.deal.list",
        "params": Object {
          "start": 27,
        },
      },
      Object {
        "method": "crm.deal.list",
        "params": Object {
          "start": 77,
        },
      },
    ],
  ],
]
`;

exports[`Bitrix \`list\` method should return a flatten payload when entities can not be fetched in a single request 1`] = `
Object {
  "error": "Expected error A
Expected error B",
  "next": 4,
  "result": Array [
    Object {
      "ID": "1",
    },
    Object {
      "ID": "2",
    },
    Object {
      "ID": "3",
    },
    Object {
      "ID": "4",
    },
  ],
  "time": Object {
    "start": 1567196890.959017,
  },
  "total": 4,
}
`;
